### PR TITLE
Feat: Adds config_change_callback to Destinations and Sources

### PR DESCRIPTION
**Pull request by tcboles**
## Summary by CodeRabbit

- **New Features**
	- Introduced an optional `config_change_callback` parameter in the `ConnectorBase`, `Destination`, and `Source` classes, enabling dynamic handling of configuration changes.
	- Enhanced the `get_destination` and `get_source` functions with the `config_change_callback` parameter to support configuration changes during setup.
	- Deprecated the `get_connector` function, encouraging users to transition to `get_source`.

- **Bug Fixes**
	- Improved error handling related to configuration change messages without affecting existing log and trace message handling.

### CODE REVIEW

**Comment by aaronsteers** [comment]
@tcboles - This looks great! I like the approach you took here - it seems nicely extensible and also simple enough to easily understand and explain to be developers.

I'm running a quick auto-fix here in CI for any simple linting or formatting issues. The slash command will only work for maintainers but you can run something similar using Ruff or Poe commands locally. 

For both "get_source()" and "get_destination()", the linter will require adding a docstring entry for the new arg. These descriptions will then feed into our auto-generated API documentation.

I'll try to provide more detailed feedback inline but for now just wanted to drop in and say "thanks!" while giving my first observations/thoughts on this implementation. Feel free to ping me here if you have other questions/blockers.

File: airbyte/_connector_base.py [hunk 1]
```diff
@@ -380,6 +383,14 @@ def _peek_airbyte_message(
                 )
             return
 
+        if (
+            message.type == "CONTROL"
+            and message.control.type == "CONNECTOR_CONFIG"
+            and self.config_change_callback is not None
+        ):
+            self.config_change_callback(message.control.config, message.control.emitted_at)
```
Comment (aaronsteers) [inline airbyte/_connector_base.py]: WDYT of only passing the new config and omitting the emitted_at input? Or do you think the developer may need both?

Comment (tcboles) [reply airbyte/_connector_base.py]: My original thought was that the developer may want to make sure the latest config was the most recent. But in theory they should always get called in order. I reverted to just sending the config.

**Comment by tcboles** [comment]
> @tcboles - This looks great! I like the approach you took here - it seems nicely extensible and also simple enough to easily understand and explain to new developers.
> 
> I'm running a quick auto-fix here in CI for any simple linting or formatting issues. The slash command will only work for maintainers but you can run something similar using Ruff or Poe commands locally.
> 
> For both "get_source()" and "get_destination()", the linter will require adding a docstring entry for the new arg. These descriptions will then feed into our auto-generated API documentation.
> 
> I'll try to provide more detailed feedback inline but for now just wanted to drop in and say "thanks!" while giving my first observations/thoughts on this implementation. Feel free to ping me here if you have other questions/blockers.


Thanks @aaronsteers. I went ahead and removed the emitted time time from the callback. I also added basic tests. Let me know if you need anything updated.

**Comment by aaronsteers** [comment]
@tcboles - Looking good! Can you remind me which source you are wanting to use this feature with, and have you had any luck testing manually with that source?

**Comment by tcboles** [comment]
> @tcboles - Looking good! Can you remind me which source you are wanting to use this feature with, and have you had any luck testing manually with that source?

I am going to be testing it with Quickbooks. I haven't had time to complete the testing yet. I will let you know when that is finished. It will hopefully be today or tomorrow.

**Comment by tcboles** [comment]
@aaronsteers Update on the testing. Everything is working for my use case with quickbooks.

**Comment by aaronsteers** [comment]
@tcboles - That's great! I'll do a final review and then merge if nothing else outstanding.

**Comment by aaronsteers** [comment]
@tcboles - Actually one more question so I can put this in the docs:

What is the shape of the data you received in the control message from QuickBooks? Can you tell me the keys/structure, obviously with anything sensitive removed?

**Comment by tcboles** [comment]
> @tcboles - Actually one more question so I can put this in the docs:
> 
> What is the shape of the config dict you received in the control message from QuickBooks? Can you tell me the keys/structure, obviously with anything sensitive removed?

Here is the full control message that was sent from source-quickbooks. The callback receives everything inside `control.connectorConfig.config`.

```
{
  "type": "CONTROL",
  "control": {
    "type": "CONNECTOR_CONFIG",
    "emitted_at": 1731090040516.7517,
    "connectorConfig": {
      "config": {
        "auth_type": "oauth2",
        "sandbox": false,
        "start_date": "2000-01-01T00:00:00Z",
        "credentials": {
          "client_id": "************",
          "client_secret": "************",
          "refresh_token": "************",
          "access_token": "************",
          "realm_id": "************",
          "token_expiry_date": "2024-11-08T19:20:40.516582+00:00"
        }
      }
    }
  }
}
```

**Review by aaronsteers — APPROVED** [review]

**Pull request merged**